### PR TITLE
Block production - eth1 data

### PR DESCRIPTION
**Pull request by mpetrunic**
- implemented algorithm for finding object with most occurrences
- required eth1 methods
- calculation of best eth1 data vote according to spec

### CODE REVIEW

**Comment by GregTheGreek** [comment]
Conflict

**Comment by mpetrunic** [comment]
> Conflict

fixd

File: src/chain/factory/block/eth1Data.ts [hunk 1]
```diff
@@ -2,40 +2,69 @@
...
+      depositCount,
+      depositRoot
+    };
+  } else {
+    const frequentVotes = mostFrequent<Eth1Data>(validVotes, Eth1Data);
+    if(frequentVotes.length === 1) {
+      return frequentVotes[0];
+    } else {
+      const blockNumbers = await Promise.all(
+        frequentVotes.map(
```
Comment (GregTheGreek) [inline src/chain/factory/block/eth1Data.ts:37]: can you clean up this map, unless it was the linter...

**Review by GregTheGreek — DISMISSED** [review]
lgtm

File: src/util/objects.ts [hunk 2]
```diff
@@ -27,3 +28,34 @@ export function isPlainObject(o: any): boolean {
...
+  count: number;
+}
+
+export function mostFrequent<T>(array: T[], type: AnySSZType): T[] {
+  const hashMap: Map<string, ElementDescription> = new Map<string, ElementDescription>();
+  array.forEach((e, index) => {
+    //We can optimize this by using faster hash like https://github.com/bevacqua/hash-sum
+    const hash = hashTreeRoot(e, type).toString('hex');
+    if(hashMap.has(hash)) {
+      const desc = hashMap.get(hash);
```
Comment (wemeetagain) [inline src/util/objects.ts]: I think it would be slightly better to do a `get` and check for null/undefined, instead of a `has` then `get`.

**Review by wemeetagain — DISMISSED** [review]

File: src/chain/factory/block/eth1Data.ts (hunk 1 above)
Comment (mpetrunic) [reply src/chain/factory/block/eth1Data.ts:37]: warning about line too long

**Comment by wemeetagain** [comment]
e2e test failed w timeout :/

**Comment by mpetrunic** [comment]
> e2e test failed w timeout :/

This is so annoying, I will add some global mocha param to increase timeout for e2e tests

**Review by GregTheGreek — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Comment by GregTheGreek** [comment]
:D

**Pull request merged**